### DISCUSSION TITLE
DateTimeTypeConverter is changing attribute name. How to prevent?

### DISCUSSION THREAD

**Discussion by RosskoDCA** [phpro/soap-client#501] [Q&A]
Hi, I struggle with this. We have soap api method, that has parameters dateFrom and dateTo and those are of type DateTime. The DateTimeTypeConverter converts this to `<date xsi:type="xsd:date">value</date>` which, of course, doesn't work.

I thought about writing custom dt-converter, but there's no option to pass original option name so I can not encode it to `dateFrom` and `dateTo`.

Can someone help me resolve this? 

Thank you

**Comment by veewee** [comment]
It's not completely clear to me what you are trying to achieve.
Could you provide some more details?

Normally ext-soap does not take into account the name of the element in the type converters. It uses the content of them instead.

**Comment by RosskoDCA** [comment]
I try my best. Basically the soap api has method getAttendanceDate(dateFrom: DateTime, dateTo: DateTime, contractNumber: string) 

I used command to generate types, classmap, client, client factory. Then I create client with ClientFactory::factory($wsdl) 

I needed to change the generated factory to use default engine factory as:
```php
    $engine = DefaultEngineFactory::create(
            $options,
            Psr18Transport::createForClient(
                new PluginClient(
                    Psr18ClientDiscovery::find(),
                    [
                        new AuthenticationPlugin(new BasicAuth('login', 'pw'))
                    ]
                )
            )
        );

```

to pass basic auth on server.

Then I try to call:
```php
$client->getAttendanceDate(new MultiArgumentRequest([
    'dateFrom' => new DateTime('2024-01-01'),
    'dateTo' => new DateTime('2024-01-31'),
    'contactNumber' => '42453453',
]);
```

which gives me result something with error (don't remember, i believe it was like `looks like we got no xml document` but i am not sure now and can't reproduce now). 
When i debug the call, there's `$request` object with XML content (prob. after using the type converter) and there I see `<date xsd:date>2024-01-01</date>` and the result from API is "dateFrom is mandatory"

When I change the call like this:
```php
$client->getAttendanceDate(new MultiArgumentRequest([
    'dateFrom' => '2024-01-01',
    'dateTo' => '2024-01-31',
    'contactNumber' => '42453453',
]);
```

the request XML contains correct `<dateFrom ...>2024-01-01</dateFrom>` but in this case the API results in "invalid type" or something.

I hope you understand what am trying to say. Am I doing something wrong?

**Reply by veewee** [reply]
Thanks for the clarification. Would it be possible to share the WSDL so that I can play around with it?

**Reply by RosskoDCA** [reply]
@veewee in attachment, is this enough for you?
[wsdl.zip](https://github.com/phpro/soap-client/files/14171120/wsdl.zip)

**Reply by veewee** [reply]
Thanks.
I've set up a little experiment and can agree this is not correct:

```php
require_once __DIR__.'/vendor/autoload.php';

use Soap\ExtSoapEngine\AbusedClient;
use Soap\ExtSoapEngine\ExtSoapEncoder;
use Soap\ExtSoapEngine\ExtSoapOptions;
use VeeWee\Xml\Dom\Document;
use function VeeWee\Xml\Dom\Configurator\pretty_print;

$client = AbusedClient::createFromOptions(ExtSoapOptions::defaults(
    'IRON_Time_And_Attendance.wsdl'
));


$encoder = new ExtSoapEncoder($client);

$request = $encoder->encode(
    'getAttendanceCheckingResults',
    [
        'dateFrom' => new \DateTimeImmutable('2024-01-01'),
        'dateTo' => new \DateTimeImmutable('2024-01-30'),
        'contractNumber' => '',
        'contractNumbers' => ['424534532'],
        'subdivisionType' => 0,
        'subdevision' => 'subdevision',
        'requiredTypeCodes' => [],
    ]
);

var_dump(Document::fromXmlString($request->getRequest(), pretty_print())->toXmlString());
```


```xml
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="urn:RON_Time_And_AttendanceIntf-IRON_Time_And_Attendance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:SOAP-ENC="http://schemas.xmlsoap.org/soap/encoding/" xmlns:ns2="urn:RON_PersonalData_Utils" xmlns:ns3="urn:RON_Time_And_AttendanceIntf" SOAP-ENV:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
  <SOAP-ENV:Body>
    <ns1:getAttendanceCheckingResults>
      <date xsi:type="xsd:date">2024-01-01</date>
      <date xsi:type="xsd:date">2024-01-30</date>
      <contractNumber xsi:type="xsd:string"/>
      <contractNumbers SOAP-ENC:arrayType="xsd:string[1]" xsi:type="ns2:TContractNumbers">
        <item xsi:type="xsd:string">424534532</item>
      </contractNumbers>
      <subdivisionType xsi:type="xsd:int">0</subdivisionType>
      <subdivision xsi:type="xsd:string">subdevision</subdivision>
      <requiredTypeCodes SOAP-ENC:arrayType="xsd:string[0]" xsi:type="ns3:TRequiredTypeCodes"/>
    </ns1:getAttendanceCheckingResults>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```


The problem here is that the `typemap` of the `ext-soap` `SoapClient` is not giving more information about what element it should use:

![Screenshot 2024-02-06 at 07 29 43](https://github.com/phpro/soap-client/assets/1618158/ea670635-da9a-47a5-86bc-7c2bff46d17a)

I'm not sure how to solve the problem at this moment.
Maybe for now, it might make more sense to disable the default type conversions inside the `ExtSoapOptions` and use a regular string instead of a `DateTimeInterface`?

**Reply by RosskoDCA** [reply]
overriding the typemap with `[]` does the trick, the request is correct with xsd:date even though i send plain string. Thanks for advice, I hope I could continue somehow now :)

**Comment by veewee** [comment]
Hello there @RosskoDCA,

I wanted to get back to this issue:
We are finishing up v4 of the soap-client package which uses a new encoding system internally.
With this new system, you should be able to get around this issue more easily.

We are eager to receive some early feedback on this new release.
If you got some time, feel free to play around with the alpha version:


* Release: https://github.com/phpro/soap-client/releases/tag/4.0.0-alpha1
* Discussion: https://github.com/phpro/soap-client/discussions/527
